server/serverAPI/controller/BaseIngredientController.ts

```diff
@@ -0,0 +1,38 @@
+import IBaseIngredient from "../model/ingredient/IBaseIngredient";
+import BaseUserController from "./BaseUserController";
+
+export default class BaseIngredientController extends BaseUserController {
+    protected sortByCategory<T extends IBaseIngredient>(collection: T[], isReverse: boolean): any {
+        let itemMap = new Map<string, T[]>();
+
+        collection.forEach(item => {
+            if (!itemMap.has(item.category)) {
+                itemMap.set(item.category, []);
+            }
+
+            itemMap.get(item.category)?.push(item);
+        });
+
+        let items = Array.from(itemMap.entries()).sort((a, b) => a[0].localeCompare(b[0]));
+
+        items.forEach(item => {
+            item[1].sort((a, b) => a.name.localeCompare(b.name))
+        });
+
+        if (isReverse) {
+            items.reverse();
+        }
+
+        return Object.fromEntries(items);
```

Comment:
For consistency, should we have all of them return the same type of response? e.g. an array?

Reply:
If u want to have this:
```
[
    [
        "catA",
        [
            {
                "id": 184270,
                "name": "B",
                "category": "catA",
                "quantityUnits": [
                    "g",
                    "kg"
                ],
                "quantity": {
                    "unit": "gas",
                    "value": 2
                },
                "recipeID": 2,
                "itemID": "6379553582a1557016d9a784"
            },
            {
                "id": 184271,
                "name": "B",
                "category": "catA",
                "quantityUnits": [
                    "g",
                    "kg"
                ],
                "quantity": {
                    "unit": "gas",
                    "value": 2
                },
                "recipeID": 2,
                "itemID": "637955f6c592584f35a64f2f"
            },
            {
                "id": 184271,
                "name": "B",
                "category": "catA",
                "quantityUnits": [
                    "g",
                    "kg"
                ],
                "quantity": {
                    "unit": "gas",
                    "value": 2
                },
                "recipeID": 1,
                "itemID": "6379561ec592584f35a64f30"
            }
        ]
    ]
]
```
Instead of this:
```
{
    "catA": [
        {
            "id": 184270,
            "name": "B",
            "category": "catA",
            "quantityUnits": [
                "g",
                "kg"
            ],
            "quantity": {
                "unit": "gas",
                "value": 2
            },
            "recipeID": 2,
            "itemID": "6379553582a1557016d9a784"
        },
        {
            "id": 184271,
            "name": "B",
            "category": "catA",
            "quantityUnits": [
                "g",
                "kg"
            ],
            "quantity": {
                "unit": "gas",
                "value": 2
            },
            "recipeID": 2,
            "itemID": "637955f6c592584f35a64f2f"
        },
        {
            "id": 184271,
            "name": "B",
            "category": "catA",
            "quantityUnits": [
                "g",
                "kg"
            ],
            "quantity": {
                "unit": "gas",
                "value": 2
            },
            "recipeID": 1,
            "itemID": "6379561ec592584f35a64f30"
        }
    ]
}
```

I can do that

Reply:
I think for consistency that would be better. Thank you